app/controllers/expenses_controller.rb

```diff
@@ -0,0 +1,36 @@
+class ExpensesController < ApplicationController
+  
+  def create
+    begin
+      create_params = params.require(:expense).permit([:name, :amount, :due_date, :debtor_id, :hideout_id, :comments])
+      expense = Expense.create!(create_params)
+      render status: 201, json: expense.to_json
+    rescue ActionController::ParameterMissing, ActiveModel::StrictValidationFailed
```

Comment:
Good catch with `ActiveModel::StrictValidationFailed`, didn't think about the fact that we would be throwing a `500` if the user provides faulty params. I left comment in `models / expense.rb` regarding the strict validation. Is there a reason we need to catch for this instead of `ActiveRecord::RecordInvalid` which is thrown when `create!` fails due to validation issue? Depending on which one we decide to use, I'll update the other controllers in my next PR.

Reply:
This is not needed in the other controllers because I see we are using `require` instead of permit for strong params. I am also taking out `permit` in favour of `require` in this controller, so this will be removed.

Reply:
Question though, say the user were to provide an integer instead of a string or vice versa for one of the attributes, does the server throw an error? If so, could you check if this is caught by `ActiveRecord::RecordInvalid`?

Reply:
Nope, that doesn't raise a db error. 
For example, when I try to save an expense with `amount  = "some string"`, the amount is type casted to 0 and is stored in the database without any errors.
I do think that it would be better to use a validation such as [`validates_numericity_of`](https://api.rubyonrails.org/classes/ActiveModel/Validations/HelperMethods.html#method-i-validates_numericality_of) in the models. However, I don't think that is high priority at the moment, we can probably leave it for later.

Reply:
Can you try doing it with a `hideout_id` in create or elsewhere that is invalid (i.e. a hideout doesn't exist with that id)? I want to know if the server will ever throw an error when we give bad record params.

Reply:
Yes, that raises a ActiveRecord::InvalidForeignKey error before rolling back.

Reply:
Do you think we should add a check for `ActiveRecord::StatementInvalid` after the other rescues so that any error that isn't a `RecordNotUnique` or `RecordNotFound` can be handled with a `400` bad request response?

Reply:
While that could for checking bad request errors I do not think we should go that path.

Mainly because `ActiveRecord::StatementInvalid` is too broad of an error. It can also potentially catch errors caused by invalid code from the controller.

For example, let say we want to get all users in a hideout. We would do something like `ActiveRecord::Base.connection.execute("SELECT * FROM users")`. But lets say we made an error in the controller and said `ActiveRecord::Base.connection.execute("SELECT * FROM user")`. This would also raise `ActiveRecord::StatementInvalid`, even though this is more of a server side error in my opinion.

I think if we wanted to validate for invalid foreign keys, invalid integer ranges, and other similar user errors, the best approach would be to use [ActiveRecord validations](https://guides.rubyonrails.org/active_record_validations.html). The rails docs also seem to recommend this approach over db side or controller solutions.

But again, this could be something to consider doing in the future. I do not think this should. be a priority right now.

Reply:
Yes, I can see how that would be an issue. However, considering that this wouldn't be applied to all controllers, only to specific ones that we have completed and therefore have caught most errors that would come from our side through various testing, I believe it would be fine a general catch **after** the main error catches have occurred. After all, we would never merge a controller that has faults like that. My reasoning for suggesting this was more so to catch the user parameter input related errors that would otherwise cause a `500 Internal Server Error`. 

That being said, how difficult would it be to implement the validations? From my review of the docs you linked, it involves modifying the `ActiveRecords` and adding the validation error `rescue` statement which is quite simple. If so, could we just proceed with that?

Reply:
Yeah, I agree that it won't cause problems if the controller implementation doesn't have faults.

`ActiveRecord` validations look pretty straightforward to implement, we should go with that in my opinion.

Reply:
Let's go with the validations then. I'll update later on my end for the remaining controllers. Lmk when its done, I'll take a look.